File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/InfoLabelWidget.java
package de.limbusdev.guardianmonsters.battle.ui.widgets;

import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.utils.Align;

import de.limbusdev.guardianmonsters.ui.Constant;

/**
 * Created by georg on 24.11.16.
 */

public class InfoLabelWidget extends de.limbusdev.guardianmonsters.battle.ui.widgets.BattleWidget
{

    private Image infoBGImg;
    public Label infoLabel;
    private String wholeText;
    private String currentText;

    public InfoLabelWidget(Skin skin) {
        super();

        infoBGImg = new Image(skin.getDrawable("label"));
        infoBGImg.setSize(372* Constant.zoom,62* Constant.zoom);
        infoBGImg.setPosition(Constant.RES_X / 2, Constant.zoom *2, Align.bottom);

        infoLabel = new Label("", skin, "default");
        infoLabel.setSize(200,58);
        infoLabel.setWrap(true);
        infoLabel.setPosition(Constant.RES_X/2, 3, Align.bottom);


        addActor(infoBGImg);
        addActor(infoLabel);
    }

    public void animateTextAppearance() {
        currentText = "";
        addAction(Actions.repeat(
            wholeText.length(),
            Actions.sequence(
                Actions.run(new Runnable() {
                @Override
                public void run() {
                    currentText += wholeText.substring(0,1);
                    wholeText = wholeText.substring(1,wholeText.length());
                    infoLabel.setText(currentText);
                }
            }),
                Actions.delay(.01f)
        )));
    }

    public void setWholeText(String wholeText)
    {
        this.wholeText = wholeText;
        clearActions();
    }
}

File: GuardianMonsters/battle/src/de/limbusdev/guardianmonsters/battle/ui/widgets/InfoLabelWidget.kt
/*
 * *************************************************************************************************
 * Copyright (c) 2019. limbusdev (Georg Eckert) - All Rights Reserved
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Georg Eckert <limbusdev.games@gmail.com>
 * *************************************************************************************************
 */

package de.limbusdev.guardianmonsters.battle.ui.widgets

import com.badlogic.gdx.scenes.scene2d.actions.Actions.*
import com.badlogic.gdx.scenes.scene2d.actions.Actions.run as runThis
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import com.badlogic.gdx.utils.Align
import de.limbusdev.guardianmonsters.battle.BattleHUD
import de.limbusdev.guardianmonsters.scene2d.*
import de.limbusdev.guardianmonsters.services.Services

import de.limbusdev.guardianmonsters.ui.Constant
import ktx.actors.then
import ktx.actors.txt
import ktx.style.get

/**
 * InfoLabelWidget displays an information label that fits into the [BattleHUD] design.
 */
open class InfoLabelWidget() : BattleWidget()
{
    // .................................................................................. Properties
    companion object { const val typingDelay = 0.01f }

    protected var infoBGImg   : Image
    private   var infoLabel   : Label

    private   var wholeText     : String = ""
    private   var remainingText : String = ""
    private   var currentText   : String = ""


    // ................................................................................ Constructors
    init
    {
        val skin = Services.UI().battleSkin

        infoBGImg = makeImage(

                drawable = skin["label"],
                layout = ImgLayout(372f * Constant.zoom, 62f, Constant.RES_Xf / 2, 2f, Align.bottom)
        )

        infoLabel = makeLabel(skin)
        infoLabel.setupLabel(LabelLayout2D(200f, 58f, Constant.RES_Xf / 2 - 100f, 61f, Align.topLeft, Align.left, true))

        addActor(infoBGImg)
        addActor(infoLabel)
    }


    // ..................................................................................... Methods
    /** Instead of showing the complete text at once, this function starts a type writer animation. */
    fun animateAsTypeWriter()
    {
        // Clear current text
        currentText = ""

        val typeWriterSequence = runThis {

            currentText += remainingText.substring(0, 1)                     // adds next character
            remainingText = remainingText.substring(1, remainingText.length) // removes first character from remaining text
            infoLabel.txt = currentText

        } then delay(typingDelay)

        addAction(repeat(wholeText.length, typeWriterSequence))
    }

    fun typeWrite(text: String)
    {
        txt = text
        animateAsTypeWriter()
    }

    /** Provides InfoLabelWidget with the complete text, the label should display. */
    var txt : String
        get() = wholeText
        set(text)
        {
            wholeText = text
            remainingText = wholeText
            clearActions()
        }
}
